packit/local_project.py

```diff
@@ -53,124 +65,175 @@ def __init__(
-                    logger.debug("it's a git repo!")
-                    change = True
-                elif self.git_url and not offline:
-                    self.git_repo = get_repo(
-                        url=self.git_url, directory=self.working_dir
-                    )
-                    logger.debug(
-                        "we just cloned git repo %s to %s",
-                        self.git_url,
-                        self.working_dir,
-                    )
-                    change = True
-
-            if (
-                self.git_url
-                and not self.working_dir
-                and not self.git_repo
-                and not offline
-            ):
-                self.git_repo = get_repo(url=self.git_url)
-                self.working_dir_temporary = True
-                change = True
-
-            if self.git_project and not self.git_url and not offline:
-                self.git_url = self.git_project.get_git_urls()["git"]
-                change = True
-
-            if self.git_project and not self.repo_name:
-                self.repo_name = self.git_project.repo
-                change = True
-
-            if self.git_project and not self.namespace:
-                self.namespace = self.git_project.namespace
-                change = True
-
-            if self.git_repo and not self.git_url:
-                # this is prone to errors
-                # also if we want url to upstream, we may want to ask for it explicitly
-                # since this can point to a fork
-                # .urls returns generator
-                self.git_url = list(self.git_repo.remote().urls)[0]
-                logger.debug(f"remote url of the repo is {self.git_url}")
-                change = True
-
-            if self.git_repo and not self.namespace:
-                self.namespace = get_namespace(self.git_url)
-                logger.debug(f"namespace from url is {self.namespace}")
-                if self.namespace:
-                    change = True
+            # we are trying to get new information while it is possible
+            # new iteration is done only if there was a change in the last iteration
+
+            change = (
+                self._parse_repo_name_full_name_and_namespace()
+                or self._parse_git_repo_from_working_dir(offline=offline)
+                or self._parse_git_project_from_repo_namespace_and_git_project(offline)
+                or self._parse_git_service_from_git_project(offline)
+                or self._parse_ref_from_git_repo()
+                or self._parse_working_dir_from_git_repo()
+                or self._parse_git_repo_from_git_url(offline)
+                or self._parse_git_url_from_git_project(offline)
+                or self._parse_repo_name_from_git_project()
+                or self._parse_namespace_from_git_project()
+                or self._parse_git_url_from_git_repo()
+                or self._parse_namespace_from_git_repo()
```

Comment:
Do we really need so many functions? Are we using all of them?

Reply:
Currently, not all of them, but I would like to compute as much as we can, so it can be useful in other projects as well.